### PR TITLE
fix calculation of alpha_lost and total_yield in case of delta<0

### PR DESCRIPTION
**Pull request by codex176743**
If delta < 0, tao_lost_bal results in a decrease in total assets.
So when calculate total_yield, consider this tao loss.

### CODE REVIEW

File: sturdy/validator/reward.py [hunk 1]
```diff
@@ -215,6 +225,10 @@ async def annualized_yield_pct(
                     bt.logging.exception(e)
             case _:
                 total_yield += 0
+        
+    if tao_lost > 0:
+        yield_loss = int(tao_lost * (apy_sum / apy_sum_count if apy_sum_count > 0 else 0))
```
[the remark this answers is not in the document]
Comment (codex176743) [reply sturdy/validator/reward.py:230]: delta<0 means that `init_alloc `is less than `pool.current_amount`. In other words, we need to unstake assets so that it results in a decrease in total assets. 
So I calculated total tao lost - `tao_lost` and we need to subtract the this amount of profit from the total_yield.
Oops, I calculated `yield_loss` wrong. Sorry. `apy_sum / apy_sum_count` was average apy when delta > 0.
Do you understand what I mean?

File: sturdy/validator/reward.py [hunk 2]
```diff
@@ -183,7 +186,9 @@ async def annualized_yield_pct(
                             _, tao_lost_bal = dynamic_info.alpha_to_tao_with_slippage(
                                 Balance.from_rao(alpha_num, netuid=pool.netuid)
                             )
-                            alpha_lost = int(tao_lost_bal.rao / (last_price / 1e9))
+                            tao_lost += tao_lost_bal.rao
+                            _, alpha_lost_bal = dynamic_info.tao_to_alpha_with_slippage(Balance.from_rao(initial_alloc))
+                            alpha_lost = alpha_lost_bal.rao
```
Comment (Shr1ftyy) [inline sturdy/validator/reward.py:191]: This doesn't make sense to me. Why do
```python
_, alpha_lost_bal = dynamic_info.tao_to_alpha_with_slippage(Balance.from_rao(initial_alloc))
```
? We're not removing that much TAO from the pool.

File: sturdy/validator/reward.py [hunk 3]
```diff
@@ -215,6 +220,9 @@ async def annualized_yield_pct(
                     bt.logging.exception(e)
             case _:
                 total_yield += 0
+        
+    if tao_lost > 0:
+        total_yield -= tao_lost
```
Comment (Shr1ftyy) [inline sturdy/validator/reward.py:225]: Straight up subtracting `tao_lost` from the `total_yield` is incorrect - it doesn't take into account the compounding effects from i.e. the staking APY. 
```python
if tao_lost > 0:
    total_yield -= tao_lost
```

**Review by Shr1ftyy — CHANGES_REQUESTED** [review]
Overall your calculations here don't really make any sense to me. 
The previous method makes much more sense imo:
```python
alpha_num = int(abs(delta) / dynamic_info.price)
_, tao_lost_bal = dynamic_info.alpha_to_tao_with_slippage(
    Balance.from_rao(alpha_num, netuid=pool.netuid)
)
alpha_lost = int(tao_lost_bal.rao / (last_price / 1e9))
...
initial_amount = int(initial_alloc / (last_price / 1e9) - alpha_lost)
```

**Comment by Shr1ftyy** [comment]
Good point. 
In that case the fix here should simply be changing:
```python
initial_amount = int(initial_alloc / (last_price / 1e9) - alpha_lost)
```
to be:
```

**Comment by Shr1ftyy** [comment]
> Okay. Let me explain in more detail. Imagine the current_amount of pool A is 100000 and allocation of a miner is 0.
> 
> ```
> current_amount = 100000
> initial_alloc = 0
> ```
> 
> `delta = initial_alloc - current_amount = -100000 < 0`
> 
> so, alpha_lost is calculated like this:
> 
> ```
> alpha_num = int(abs(delta) / dynamic_info.price)
> _, tao_lost_bal = dynamic_info.alpha_to_tao_with_slippage(
>     Balance.from_rao(alpha_num, netuid=pool.netuid)
> )
> alpha_lost = int(tao_lost_bal.rao / (last_price / 1e9))
> ```
> 
> then,
> 
> ```
> initial_amount = int(initial_alloc / (last_price / 1e9) - alpha_lost)
> alpha_amount = int((initial_amount) * (1 + annualized_alpha_apy))
> ```
> 
> where `initial_alloc =0`, so `initial_amount = -alpha_lost` is it right?
> 
> because inital_alloc is 0, alpha_amount should be 0.
> 
> what do you think of it?

Good point. 
In this case the fix should simply be changing the following:
```python
initial_amount = int(initial_alloc / (last_price / 1e9) - alpha_lost)
```
to instead be:
```python
initial_amount_raw = int(initial_alloc / (last_price / 1e9) - alpha_lost)
initial_amount = initial_amount_raw if initial_amount_raw >= 0 else 0
```

**Comment by Shr1ftyy** [comment]
> Okay. Let me explain in more detail. Imagine the current_amount of pool A is 100000 and allocation of a miner is 0.
> 
> ```
> current_amount = 100000
> initial_alloc = 0
> ```
> 
> `delta = initial_alloc - current_amount = -100000 < 0`
> 
> so, alpha_lost is calculated like this:
> 
> ```
> alpha_num = int(abs(delta) / dynamic_info.price)
> _, tao_lost_bal = dynamic_info.alpha_to_tao_with_slippage(
>     Balance.from_rao(alpha_num, netuid=pool.netuid)
> )
> alpha_lost = int(tao_lost_bal.rao / (last_price / 1e9))
> ```
> 
> then,
> 
> ```
> initial_amount = int(initial_alloc / (last_price / 1e9) - alpha_lost)
> alpha_amount = int((initial_amount) * (1 + annualized_alpha_apy))
> ```
> 
> where `initial_alloc =0`, so `initial_amount = -alpha_lost` is it right?
> 
> because inital_alloc is 0, alpha_amount should be 0.
> 
> what do you think of it?

Good point. 
In this case the fix should simply be changing the following:
```python
initial_amount = int(initial_alloc / (last_price / 1e9) - alpha_lost)
alpha_amount = int((initial_amount) * (1 + annualized_alpha_apy))
tao_pct_return = ((alpha_amount * (curr_price / 1e9)) - (initial_alloc)) / (initial_alloc)
```
to instead be:
```python
initial_amount_raw = int(initial_alloc / (last_price / 1e9) - alpha_lost)
if initial_amount_raw >= 0:
  initial_amount = initial_amount_raw
  alpha_amount = int((initial_amount) * (1 + annualized_alpha_apy))
  tao_pct_return = ((alpha_amount * (curr_price / 1e9)) - (initial_alloc)) / (initial_alloc)
else:
  initial_amount = 0 # if initial_amount raw is less than zero than there is no initial amount
  alpha_amount = 0 # same can be said for the alpha amount
  tao_pct_return = (alpha_lost * (last_price / 1e9)) - (initial_alloc)) / (initial_alloc) # but we still want to consider the % loss from slippage - which is done here.
```

@romanorac @codex176743 would love some thoughts on this.

**Comment by Shr1ftyy** [comment]
please see my updated comment above. I believe you are responding to an older comment I deleted @codex176743

**Comment by Shr1ftyy** [comment]
>because we allocate initial_alloc, it should be like this.
```
_, alpha_lost_bal = dynamic_info.tao_to_alpha_with_slippage(Balance.from_rao(initial_alloc))
alpha_lost = alpha_lost_bal.rao
```

Also this is incorrect because we are not actually moving `initial_alloc` into the pool, but rather moving `delta` tao into the pool, which is why we perform the following calculation:
```python
elif delta < 0:
    alpha_num = int(abs(delta) / dynamic_info.price)
    _, tao_lost_bal = dynamic_info.alpha_to_tao_with_slippage(
        Balance.from_rao(alpha_num, netuid=pool.netuid)
    )
    alpha_lost = int(tao_lost_bal.rao / (last_price / 1e9))

```

**Comment by codex176743** [comment]
you're right

**Comment by codex176743** [comment]
But i don't understand this;
`tao_pct_return = (alpha_lost * (last_price / 1e9)) - (initial_alloc)) / (initial_alloc)`

**Comment by codex176743** [comment]
`tao_pct_return = (alpha_lost * (last_price / 1e9) - (initial_alloc)) / (initial_alloc)`

where `alpha_lost * (last_price / 1e9) > initial_alloc`, so tao_pct_return > 0.
is it right?

**Comment by codex176743** [comment]
>   initial_amount = 0 # if initial_amount raw is less than zero than there is no initial amount
>   alpha_amount = 0 # same can be said for the alpha amount
>   tao_pct_return = (alpha_lost * (last_price / 1e9)) - (initial_alloc)) / (initial_alloc) # but we still want to consider the % loss from slippage - which is done here.

you wrote: `tao_pct_return = (alpha_lost * (last_price / 1e9)) - (initial_alloc)) / (initial_alloc) `
should be: `tao_pct_return = (alpha_lost * (last_price / 1e9) - (initial_alloc)) / (initial_alloc)`

**Comment by Shr1ftyy** [comment]
> But i don't understand this; `tao_pct_return = (alpha_lost * (last_price / 1e9)) - (initial_alloc)) / (initial_alloc)`

Wait, my bad, this is wrong. If we want to consider the amount of tao lost from slippage what we should be doing is:
```python
initial_amount_raw = int(initial_alloc / (last_price / 1e9) - alpha_lost)
if initial_amount_raw >= 0:
  initial_amount = initial_amount_raw
  alpha_amount = int((initial_amount) * (1 + annualized_alpha_apy))
  tao_pct_return = ((alpha_amount * (curr_price / 1e9)) - (initial_alloc)) / (initial_alloc)
  total_yield += int(tao_pct_return * initial_alloc)
else:
  initial_amount = 0 # if initial_amount raw is less than zero than there is no initial amount
  alpha_amount = 0 # same can be said for the alpha amount
  tao_delta = alpha_lost * (last_price / 1e9) # but we still want to consider the % loss from slippage - which is done here.
  total_yield += tao_delta
```

**Comment by codex176743** [comment]
That's it.

```
if tao_lost > 0:
        total_yield -= tao_lost
```
This is what I want to say.

**Comment by Shr1ftyy** [comment]
right. I think the surrounding changes you made don't make sense. if you make the changes I suggested above instead, I think we'd be good to merge this PR.

**Comment by codex176743** [comment]
Okay.

File: sturdy/validator/reward.py [hunk 4]
```diff
@@ -171,7 +171,7 @@ async def annualized_yield_pct(
                         delta = initial_alloc - pool.current_amount
 
                         # alpha delta (in tao)
-                        alpha_delta_tao = delta / (last_price + 1)
+                        alpha_delta_tao = delta / last_price
```
Comment (Shr1ftyy) [inline sturdy/validator/reward.py:174]: I think we just keep the 1 here imo.

**Review by Shr1ftyy — CHANGES_REQUESTED** [review]
smol nitpick

File: sturdy/validator/reward.py (hunk 4 above)
Comment (codex176743) [reply sturdy/validator/reward.py:174]: Okay

**Comment by Shr1ftyy** [comment]
nice work @codex176743.
one more thing - can you please bump the version from 3.0.14 to 3.0.15? See the following commit to know where to do this: https://github.com/Sturdy-Subnet/sturdy-subnet/pull/118/commits/9d356f2ab0c85fec16e0565c2788c281cb3504c8

**Comment by Shr1ftyy** [comment]
Nice work.
Merging.

**Comment by codex176743** [comment]
Great. Thanks.
